dbms/src/Storages/DeltaMerge/ReadThread/ColumnSharingCache.h

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include <Columns/IColumn.h>
+#include <DataTypes/IDataType.h>
+#include <Storages/DeltaMerge/File/DMFile.h>
+
+#include <memory>
+
+namespace DB::DM
+{
+enum class ColumnCacheStatus
+{
+    ADD_COUNT = 0,
+    ADD_STALE,
+    GET_MISS,
+    GET_PART,
+    GET_HIT,
+    GET_COPY,
+
+    _TOTAL_COUNT,
+};
+
+class ColumnSharingCache
+{
+public:
+    struct ColumnData
+    {
+        size_t pack_count{0};
+        ColumnPtr col_data;
+    };
+
+    void add(size_t start_pack_id, size_t pack_count, ColumnPtr & col_data)
+    {
+        std::lock_guard lock(mtx);
+        auto & value = packs[start_pack_id];
+        if (value.pack_count < pack_count)
+        {
+            value.pack_count = pack_count;
+            value.col_data = col_data;
+        }
+    }
+
+    ColumnCacheStatus get(size_t start_pack_id, size_t pack_count, size_t read_rows, ColumnPtr & col_data, DataTypePtr data_type)
+    {
+        ColumnCacheStatus status;
+        std::lock_guard lock(mtx);
+        auto target = packs.find(start_pack_id);
+        if (target == packs.end())
+        {
+            status = ColumnCacheStatus::GET_MISS;
+        }
+        else if (target->second.pack_count < pack_count)
+        {
+            status = ColumnCacheStatus::GET_PART;
```

Comment:
I think there is space to optimize this part further, i.e. even if part of the cache that exists can still be helpful.
But we can do it later.

Reply:
And to reduce the chance of step mismatch of the split of packs, we can try to align the split point. For example, we always read packs by 4 steps.  

Thread 1:         2, 3 | 4, 5, 6, 7 | 8, 9
Thread 2: 0, 1, 2, 3 | 4, 5, 6, 7 | 8,